exp/api/v1alpha3/conversion.go

```diff
@@ -17,26 +17,72 @@ limitations under the License.
 package v1alpha3
 
 import (
-	"k8s.io/apimachinery/pkg/conversion"
-	v1beta1 "sigs.k8s.io/cluster-api/exp/api/v1beta1"
+	apimachineryconversion "k8s.io/apimachinery/pkg/conversion"
+	"sigs.k8s.io/cluster-api/exp/api/v1beta1"
+	utilconversion "sigs.k8s.io/cluster-api/util/conversion"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
 )
 
 // Convert_v1alpha3_MachinePoolSpec_To_v1beta1_MachinePoolSpec is an autogenerated conversion function.
-func Convert_v1alpha3_MachinePoolSpec_To_v1beta1_MachinePoolSpec(in *MachinePoolSpec, out *v1beta1.MachinePoolSpec, s conversion.Scope) error {
+func Convert_v1alpha3_MachinePoolSpec_To_v1beta1_MachinePoolSpec(in *MachinePoolSpec, out *v1beta1.MachinePoolSpec, s apimachineryconversion.Scope) error {
 	return autoConvert_v1alpha3_MachinePoolSpec_To_v1beta1_MachinePoolSpec(in, out, s)
 }
 
-func Convert_v1alpha3_MachinePool_To_v1beta1_MachinePool(in *MachinePool, out *v1beta1.MachinePool, s conversion.Scope) error {
+func Convert_v1alpha3_MachinePool_To_v1beta1_MachinePool(in *MachinePool, out *v1beta1.MachinePool, s apimachineryconversion.Scope) error {
 	if err := autoConvert_v1alpha3_MachinePool_To_v1beta1_MachinePool(in, out, s); err != nil {
 		return err
 	}
 
 	// Replace v1alpha3 finalizer to allow old MachinePools to get deleted.
-	if !controllerutil.ContainsFinalizer(out, MachinePoolFinalizer) {
+	if controllerutil.ContainsFinalizer(out, MachinePoolFinalizer) {
```

Comment:
Do we need to do the reverse operation when converting from v1beta1 to v1alpha3
@CecileRobertMichon to confirm about this and the changes in the expression

Reply:
yes, I think so.. In practice this won't be used much (I don't foresee many users doing beta to alpha3 conversion for MachinePool and manually changing the group) but we should cover it.

Reply:
thanks for fixing the logic, that looks correct to me

Reply:
Should be fixed now, PTAL